### PR TITLE
s4548663-GFNet-ADNI

### PR DESCRIPTION
**Pull request by YifeiZhang233eV**
### Description
Dear 3710 team, this PR introduces and builds the GFNet model to classify brain MRI scans as either Alzheimer's Disease (AD) or Cognitive Normal (CN), using the ADNI dataset. Significant improvements have been made to ensure model accuracy, robustness, and memory efficiency.

### Implementation
1. dataset.py: Load and preprocess the ADNI data within PyTorch custom data loader;
2. modules.py: Build a GFNet model structure;
3. train.py: Train a best model using the structure in modules.py by the data in the train set;
4. predict.py: Get the result of the test dataset by the trained model;
5. README.md: Including the description of the dataset and model structure, the final result with figures, future work if possible, and dependencies.

### Testing
1. Run the train.py to get the trained model in ./models;
2. Run the predict.py to the accuracy result of the test set.

### Notes:
1. Increased training time: This may lead to increased training time due to the use of a large number of data augmentation methods and small batch sizes.
2. Memory requirements: Although memory usage is controlled by reducing the batch size, the model still requires large GPU memory. If running on a low-profile device, you may need to adjust the batch size further.
3. Model complexity: Using multi-layer Transformer blocks and MLPS may make the model more complex, and the generalization ability on different datasets needs further validation.

### Others
I submitted this homework late because I applied for an extension. I'm sorry for everyone's extra workload. Thank you for your comments. I will make sure to correct any problems in time.

### CODE REVIEW

**Comment by wangzhaomxy** [comment]
**<This is an initial inspection, no action is required at this point.>**

**File Organizing:** Well-organized files.

**Problem Solving:**

- The algorithm solves the problem appropriately. 
- Accuracy in testing dataset: 0.7485. 

**Model and functions:**

- It correctly uses PyTorch to construct the GFNet models and functions. 
- With data augmentation.
- Training code cannot work as the model is initialized as input channel = 3 while the input image channel is 1, so an input shape error will be posted when the images are fed into the model.

**Code design:** Most of the functions are good other than the input channel error leading to code running failure.

**Code comment and docstring:**

- Good code comments 
- Minmal function docstrings

**Difficulty:** Hard.

**Additional Comments:**

- Reasonable commits
- Good ReadMe design

**Comment by aniketgupta17** [comment]
Observational Feedback

Pull Request:
Correctly created the Pull request from Topic Recognition Branch .
The pull request includes a clear description about the file structure .


File Organizing: Well-organized files.

Commit Log:
Commit messages are  progressive for the Recognition Problem solved using 4 files .
Commits are regularly made, showing incremental development rather than bulk updates.

Documentation:
The README file could be made more readable with use of bullet points and organised code blocks . 
Code comments are included.
Can add about Conclusion and future improvements in the ReadMe.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Accuracy in testing dataset: 0.7485.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting missing. Minimal doc strings</td><td>-1</td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. GFNet (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>Feedback action required: Feedback marks possible +2 if the requested changes are made. Remove model checkpoint, cache files.</td><td>-2</td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-5</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Approved extension +2

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files (model files). Please update for merge, doesn't affect grade.
